### PR TITLE
Handle setopts results instead of crashing

### PR DESCRIPTION
**Pull request by bjosv**
- If setting socket options during a connection procedure fails
we now return the error instead of causing a crash.
This was the last known crash-point when starting a client.

- If setting the socket active again after receiving some data
we now close the socket and reconnect (..if reconnection not disabled)

### CODE REVIEW

File: src/eredis_client.erl [hunk 1]
```diff
@@ -372,7 +375,7 @@ connect_next_addr([Addr|Addrs], Port, SocketOptions, ConnectOptions, State) ->
                                     {ok, State#state{socket = NewSocket}};
                                 {error, Reason} ->
                                     close_socket(State, NewSocket),
-                                    {error, {select_error, Reason}}
+                                    {error, {select_db_error, Reason}}
```
Comment (zuiderkwast) [inline src/eredis_client.erl:378]: Why change this? Maybe someone relies on this atom...

**Review by zuiderkwast — COMMENTED** [review]
I believe start_link can still cause an exit signal to be emitted, because init can return `{stop, Reason}`. If Reason is anything other than `normal`, the process will emit an exit signal.

I think this is acceptable though, if reconnect is disabled.

File: src/eredis_client.erl (hunk 1 above)
Comment (bjosv) [reply src/eredis_client.erl:378]: Just niceness, but you are right, Ill revert that

File: src/eredis_client.erl [hunk 2]
```diff
@@ -143,15 +143,18 @@ handle_cast(_Msg, State) ->
...
-%% Receive TLS data from socket. Match `Socket' to enforce sanity.
-handle_info({ssl, Socket, Data}, #state{socket = Socket} = State) ->
-    ok = ssl:setopts(Socket, [{active, once}]),
-    {noreply, handle_response(Data, State)};
+%% Receive TCP/TLS data from socket. Match `Socket' to enforce sanity.
+handle_info({Type, Socket, Data},
+            #state{socket = Socket, transport = Transport} = State1)
+  when Type =:= tcp orelse Type =:= ssl ->
+    State = handle_response(Data, State1),
+    case setopts(Socket, Transport, [{active, once}]) of
```
Comment (bjosv) [inline src/eredis_client.erl:151]: Now there is a dynamic call here, how bad is it performance wise?

**Review by zuiderkwast — APPROVED** [review]

**Pull request merged**